src/coreclr/jit/fgehopt.cpp

```diff
@@ -2097,3 +2097,649 @@ PhaseStatus Compiler::fgTailMergeThrows()
+        {
+            EHblkDsc* const enclosingTryEbd = ehGetDsc(enclosingTryIndex);
+
+            if (EHblkDsc::ebdIsSameTry(ebd, enclosingTryEbd))
+            {
+                JITDUMP("==> found mutual-protect try EH#%02u for EH#%02u\n", enclosingTryIndex, regionIndex);
+                regionsToProcess.Push(enclosingTryIndex);
+            }
+        }
+
+        JITDUMP("<== finished try EH#%02u\n", regionIndex);
+    }
+
+    // Find the outermost mututal-protect try egion that begins at tryEntry
+    //
+    EHblkDsc* const tryEbd            = ehGetDsc(tryIndex);
+    unsigned        outermostTryIndex = tryIndex;
+    unsigned        enclosingTryIndex = EHblkDsc::NO_ENCLOSING_INDEX;
+    {
+        EHblkDsc* outermostEbd = ehGetDsc(outermostTryIndex);
+        while (true)
+        {
+            enclosingTryIndex = outermostEbd->ebdEnclosingTryIndex;
+            if (enclosingTryIndex == EHblkDsc::NO_ENCLOSING_INDEX)
+            {
+                break;
+            }
+            outermostEbd = ehGetDsc(enclosingTryIndex);
+            if (!EHblkDsc::ebdIsSameILTry(outermostEbd, tryEbd))
+            {
+                break;
+            }
+            outermostTryIndex = enclosingTryIndex;
+        }
+    }
+
+    unsigned enclosingHndIndex = EHblkDsc::NO_ENCLOSING_INDEX;
+    if (tryEntry->hasHndIndex())
+    {
+        enclosingHndIndex = tryEntry->getHndIndex();
+    }
+
+    // Now blocks contains an entry for each block to clone,
+    // and maps (if non-null) contains a key for each block to clone, with nullptr value
+    //
+    JITDUMP("Will need to clone %u EH regions (outermost: EH#%02u) and %u blocks\n", regionCount, outermostTryIndex,
+            blocks.size());
+
+    // Allocate the new EH clauses. First, find the enclosing EH clause, if any...
+    // we will want to allocate the new clauses just "before" this point.
+    //
+    // If the region we're cloning is not enclosed, we put it at the end of the table;
+    // this is cheaper than any other insertion point, as no existing regions get renumbered.
+    //
+    unsigned insertBeforeIndex = enclosingTryIndex;
+    if (insertBeforeIndex == EHblkDsc::NO_ENCLOSING_INDEX)
+    {
+        JITDUMP("Cloned EH clauses will go at the end of the EH table\n");
+        insertBeforeIndex = compHndBBtabCount;
+    }
+    else
+    {
+        JITDUMP("Cloned EH clauses will go before enclosing region EH#%02u\n", enclosingTryIndex);
+    }
+
+    // Once we call fgAddEHTableEntries, all the EH indicies at or above insertBeforeIndex
+    // will shift, and the EH table may reallocate.
+    //
+    // This addition may also fail, if the table would become too large...
+    //
+    EHblkDsc* const clonedOutermostEbd =
+        fgAddEHTableEntries(insertBeforeIndex, regionCount, /* deferAdding */ deferCloning);
+
+    if (clonedOutermostEbd == nullptr)
+    {
+        JITDUMP("fgCloneTryRegion: unable to expand EH table\n");
+        return nullptr;
+    }
```

Comment:
Maybe rename this function to `fgTryAddEHTableEntries` if it is expected that it can fail?